lib/puppet/util/windows/service.rb

```diff
@@ -0,0 +1,726 @@
+                resume_ptr.write_int(0)
+                if EnumServicesStatusExW(
+                  scm,
+                  SC_ENUM_PROCESS_INFO,
+                  ALL_SERVICE_TYPES,
+                  SERVICE_STATE_ALL,
+                  buffer_ptr,
+                  buffer_ptr.size,
+                  bytes_pointer,
+                  svcs_ret_ptr,
+                  resume_ptr,
+                  nil,
+                ) == FFI::WIN32_FALSE
+                  raise Puppet::Util::Windows::Error.new(_("Failed to fetch services"))
+                end
+                # Now that the buffer is populated with services
+                # we pull the data from memory using pointer arithmetic:
+                # the number of services returned by the function is
+                # available to be read from svcs_ret_ptr, and we iterate
+                # that many times moving the cursor pointer the length of
+                # ENUM_SERVICE_STATUS_PROCESSW.size. This should iterate
+                # over the buffer and extract each struct.
+                services_returned = svcs_ret_ptr.read(:dword)
+                cursor_ptr = buffer_ptr
+                1.upto(services_returned) do |index|
+                  service = ENUM_SERVICE_STATUS_PROCESSW.new(cursor_ptr)
+                  services[service[:lpServiceName].read_arbitrary_wide_string_up_to] = {
+                    'display_name' => service[:lpDisplayName].read_arbitrary_wide_string_up_to,
+                    'service_status_process' => service[:ServiceStatusProcess]
+                  }
+                  cursor_ptr += ENUM_SERVICE_STATUS_PROCESSW.size
+                end
+              end # resume_ptr
+            end # svcs_ret_ptr
+          end # buffer_ptr
+        end # bytes_ptr
+      end # open_scm
+      services
+    end
+    module_function :enum_services
+
+    class << self
+      # @api private
+      # Opens a connection to the SCManager on windows then uses that
+      # handle to create a handle to a specific service in windows
+      # corresponding to service_name
+      #
+      # this function takes a block that executes within the context of
+      # the open service handler, and will close the service and SCManager
+      # handles once the block finishes
+      #
+      # @param [string] service_name the name of the service to open
+      # @param [Integer] scm_access code corresponding to the access type requested for the scm
+      # @param [Integer] service_access code corresponding to the access type requested for the service
+      # @yieldparam [:handle] service the windows native handle used to access
+      #   the service
+      def open_service(service_name, scm_access, service_access)
+        service = FFI::Pointer::NULL_HANDLE
+        open_scm(scm_access) do |scm|
+          FFI::MemoryPointer.from_string_to_wide_string(service_name) do |service_name_ptr|
+            service = OpenServiceW(scm, service_name_ptr, service_access)
```

Comment:
You could avoid this block by just using `wide_string` like here: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/windows/security.rb#L170

Reply:
Oooooo, yeah, I think you're right since it's type is LPCWSTR. For some reason I thought it was not a constant string